libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c

```diff
@@ -1235,23 +1235,6 @@ CK_DEFINE_FUNCTION( CK_RV, C_Finalize )( CK_VOID_PTR pvReserved )
         {
             vSemaphoreDelete( xP11Context.xObjectList.xMutex );
             memset( &xP11Context, 0, sizeof( xP11Context ) );
-
-            /* Close device private key which is a persistent key. Note that here we only close it other than
-             * destroying it so that the key can be activated by calling psa_open_key then next time it is used.
-             */
-            xResult = PKCS11PSARemoveObject( (uint8_t *)pkcs11configLABEL_DEVICE_PRIVATE_KEY_FOR_TLS,
```

Comment:
the decision can be made by the upper layer. If the object no longer needed, it should call C_DestroyObject() before close the session. 

In AFR IDT testing, it has such scenario 
 1. Create key/certificate objects 
 2. run test_case_1 
 3. run test_case_2
 4. Destroy the objects. 

if the object is removed while close the session 1, the  test_case_2 will fail as objects can't be found.

Reply:
If the object is removed and closed in C_Finalize, then in the c_findobjects, it will open the key and import it into the P11KeyConfig. So in test_case_2, it will find the object by psa_open_key. This is my implementation.
If does not remove the objects here, and does not call PKCS11PSAGetKeyHandle in c_findobjects, the key will be opened multi-times during the tests while it is already opened. This may cause open key failure as there are only 16 key entries in TF-M. This will cause the test_case_2 fail.
In your code, PKCS11PSAGetKeyHandle is called in c_findobjects and the key is not removed here, so the key is only opened one time. This should also work.

Reply:
In your implementation, the key is not closed in the PKCS#11 and uses the same key_handle that is returned by psa_open_key. But the key may be closed by other APPs, if so, the original key handle is not valid anymore or it has been allocated to another key. So I prefer to closing the key after using it.